{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process all file in folder\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List the files to process in both conditions wt and sega"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files 62\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "# list files from the src folder\n",
    "src = Path('/media/cephfs2/jparham/Joe for Jerome /')\n",
    "dst = Path('./')\n",
    "\n",
    "filelist = pd.concat((\n",
    "    pd.DataFrame.from_records(\n",
    "        [{'path':f, 'name':f.name, 'condition':'sega'}  for d in src.glob('*/*/Crop*') for f in d.glob('[!.]*.tif')]\n",
    "    ),\n",
    "    pd.DataFrame.from_records(\n",
    "    [{'path':f, 'name':f.name, 'condition':'wt'}  for d in src.glob('Div*') for f in d.glob('[!.]*.tif')]\n",
    ")),axis=0,ignore_index=True)\n",
    "\n",
    "print(f\"Number of files {len(filelist)}\")\n",
    "filelist.head()\n",
    "filelist.to_csv('filelist.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process all file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from tqdm.notebook import tqdm\n",
    "# tqdm.pandas()\n",
    "# import motionquant as mq\n",
    "\n",
    "# results_path = dst/Path('results.h5') # result\n",
    "# if results_path.exists():\n",
    "#     results_path.unlink()\n",
    "\n",
    "# print(f\"Saving results in {results_path}\")\n",
    "\n",
    "# df = []\n",
    "# for filename in tqdm(filelist['path'].iloc):\n",
    "#     try:\n",
    "#         name = filename.stem        \n",
    "#         img, mask, position, speed, diff, flow, rho, div, blob = mq.process(filename)\n",
    "#         df.append(mq.record(name, img, mask, position, speed, diff, flow, rho, div, blob))\n",
    "#         mq.save_result(results_path, name, img, mask, position, speed, diff, flow, rho, div, blob)\n",
    "#     except Exception as e:\n",
    "#         print(f\"file '{filename}' could not be processed\")\n",
    "#         print(e)\n",
    "#         pass\n",
    "# df = pd.concat(df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parallel processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8dcfdb6623dd46868401f89ae6acc5f0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<div class=\"jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output\">\n",
       "    <div style=\"width: 24px; height: 24px; background-color: #e1e1e1; border: 3px solid #9D9D9D; border-radius: 5px; position: absolute;\">\n",
       "    </div>\n",
       "    <div style=\"margin-left: 48px;\">\n",
       "        <h3 style=\"margin-bottom: 0px; margin-top: 0px;\">LocalCluster</h3>\n",
       "        <p style=\"color: #9D9D9D; margin-bottom: 0px;\">932d31c9</p>\n",
       "        <table style=\"width: 100%; text-align: left;\">\n",
       "            <tr>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Dashboard:</strong> <a href=\"http://127.0.0.1:8787/status\" target=\"_blank\">http://127.0.0.1:8787/status</a>\n",
       "                </td>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Workers:</strong> 4\n",
       "                </td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Total threads:</strong> 12\n",
       "                </td>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Total memory:</strong> 31.27 GiB\n",
       "                </td>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "    <td style=\"text-align: left;\"><strong>Status:</strong> running</td>\n",
       "    <td style=\"text-align: left;\"><strong>Using processes:</strong> True</td>\n",
       "</tr>\n",
       "\n",
       "            \n",
       "        </table>\n",
       "\n",
       "        <details>\n",
       "            <summary style=\"margin-bottom: 20px;\">\n",
       "                <h3 style=\"display: inline;\">Scheduler Info</h3>\n",
       "            </summary>\n",
       "\n",
       "            <div style=\"\">\n",
       "    <div>\n",
       "        <div style=\"width: 24px; height: 24px; background-color: #FFF7E5; border: 3px solid #FF6132; border-radius: 5px; position: absolute;\"> </div>\n",
       "        <div style=\"margin-left: 48px;\">\n",
       "            <h3 style=\"margin-bottom: 0px;\">Scheduler</h3>\n",
       "            <p style=\"color: #9D9D9D; margin-bottom: 0px;\">Scheduler-c1f49571-756f-4399-a9a0-32cd6c653591</p>\n",
       "            <table style=\"width: 100%; text-align: left;\">\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Comm:</strong> tcp://127.0.0.1:34851\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Workers:</strong> 4\n",
       "                    </td>\n",
       "                </tr>\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Dashboard:</strong> <a href=\"http://127.0.0.1:8787/status\" target=\"_blank\">http://127.0.0.1:8787/status</a>\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Total threads:</strong> 12\n",
       "                    </td>\n",
       "                </tr>\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Started:</strong> Just now\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Total memory:</strong> 31.27 GiB\n",
       "                    </td>\n",
       "                </tr>\n",
       "            </table>\n",
       "        </div>\n",
       "    </div>\n",
       "\n",
       "    <details style=\"margin-left: 48px;\">\n",
       "        <summary style=\"margin-bottom: 20px;\">\n",
       "            <h3 style=\"display: inline;\">Workers</h3>\n",
       "        </summary>\n",
       "\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 0</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:42401\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 3\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:43903/status\" target=\"_blank\">http://127.0.0.1:43903/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 7.82 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:37017\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /tmp/dask-scratch-space/worker-m43ma5k7\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 1</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:36401\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 3\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:41135/status\" target=\"_blank\">http://127.0.0.1:41135/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 7.82 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:43793\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /tmp/dask-scratch-space/worker-03r5q_nu\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 2</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:34033\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 3\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:34721/status\" target=\"_blank\">http://127.0.0.1:34721/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 7.82 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:35659\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /tmp/dask-scratch-space/worker-h49k_pwl\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 3</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:40299\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 3\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:38509/status\" target=\"_blank\">http://127.0.0.1:38509/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 7.82 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:45223\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /tmp/dask-scratch-space/worker-8488wai0\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "\n",
       "    </details>\n",
       "</div>\n",
       "\n",
       "        </details>\n",
       "    </div>\n",
       "</div>"
      ],
      "text/plain": [
       "LocalCluster(932d31c9, 'tcp://127.0.0.1:34851', workers=4, threads=12, memory=31.27 GiB)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from dask.distributed import LocalCluster\n",
    "cluster = LocalCluster()          # Fully-featured local Dask cluster\n",
    "client = cluster.get_client()\n",
    "cluster.scale(6)\n",
    "cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a58a45\n"
     ]
    }
   ],
   "source": [
    "import git\n",
    "repo = git.Repo(search_parent_directories=True)\n",
    "tag = repo.head.object.hexsha[-6:]\n",
    "print(tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving results in file 'results.h5'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "WARNING: no mask pixels found\n",
      "WARNING: no mask pixels found\n",
      "WARNING: no mask pixels found\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n",
      "/home/jeromeb/miniconda3/envs/imaging/lib/python3.9/site-packages/skimage/measure/_regionprops.py:594: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  tuple(M[_get_element(axis)] / M0 for axis in range(self._ndim)))\n"
     ]
    }
   ],
   "source": [
    "import motionquant as mq\n",
    "import dask\n",
    "from dask.distributed import Lock\n",
    "import torch\n",
    "import traceback\n",
    "\n",
    "def process(filename, results_path):\n",
    "    name = filename.stem        \n",
    "    img, mask, position, speed, diff, flow, rho, div, blob = mq.process(filename)\n",
    "    lock = Lock('process-sufo')\n",
    "    lock.acquire()\n",
    "    mq.save_result(results_path, name, img, mask, position, speed, diff, flow, rho, div, blob)\n",
    "    lock.release()\n",
    "    df = mq.record(name, img, mask, position, speed, diff, flow, rho, div, blob)\n",
    "    return df\n",
    "\n",
    "def process_safe(filename, results_path):\n",
    "    try:\n",
    "        return process(filename, results_path)\n",
    "    except Exception as e:\n",
    "        print(f\"file '{filename}' could not be processed\")\n",
    "        print(e)\n",
    "        print(traceback.print_exc())\n",
    "        pass\n",
    "\n",
    "\n",
    "results_path = dst/Path(f'results-{tag}.h5') # result\n",
    "if results_path.exists():\n",
    "    results_path.unlink()\n",
    "\n",
    "print(f\"Saving results in file '{results_path}'\")\n",
    "tsk = [dask.delayed(process_safe)(filename, results_path) for filename in filelist['path'].iloc]\n",
    "res = dask.compute(tsk)\n",
    "df = pd.concat(res[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(dst/'results.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "imaging",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
